Add System Simulations

mitchute opened the pull request:
## Pull request overview

PR adds system simulation capabilities. This is based on work by @niranjan-bas and @jeffreyspitler.

### Checklist

- [ ] Documentation added/updated
- [x] CI status: all green or justified

mitchute committed:
initial version of system simulaton code

Myoldmopar reviewed (commented):
I know you've made tons of cleanups here already, and it looks good.  I made a few comments, but take them or leave them as you please.  

If the custom text input format is going to be actually included as a part of GHEDesigner, there will need to be gobs and gobs of additional error handling and testing though.  :-|

mitchute commented on ghedesigner/district_system.py:
```diff
@@ -0,0 +1,621 @@
+
+        # Thermal object references (to be set during setup)
+        self.pipe = Pipe
+        self.soil = Soil
```
Yep, makes sense. I'll see what I can do.

mitchute commented on ghedesigner/district_system.py:
```diff
@@ -0,0 +1,621 @@
+
+    def read_ghe_hp_system_data(self, data, data_dir: Path):
+        next_matrix_line = 0
+        for line in data:  # loop over all the lines
```
No, we're moving it all over to JSON, but will also likely add the ability for the loads to reference one or more csv files.

mitchute committed:
add new example loads file

mitchute committed:
Merge branch 'develop' into implement-sys-simulation

mitchute committed:
remove unused hydralic-run flag, start migrating over to json input data

mitchute committed:
remove matrix line

mitchute committed:
move ghe over to json data file

mitchute committed:
add output file to tests dir for simpler testing

mitchute committed:
start pulling in hp data from json

mitchute committed:
cleanups

mitchute committed:
cleaning up constructor

mitchute committed:
move pipe to json data

mitchute committed:
move grout over to json data

mitchute committed:
move soil to json data

mitchute committed:
move borehole to json data

mitchute committed:
moving to ghe ctor

mitchute committed:
moving to ghe ctor

mitchute committed:
fix - move gfunction and c_n to be instances on each GHE, as was intended

mitchute committed:
moving things to proper ctors

mitchute committed:
cleanups

mitchute committed:
cleanups

mitchute committed:
remove buildings

mitchute committed:
move from zones to buildings

mitchute committed:
start transitioning buildings to json data

mitchute committed:
continuing to transition buildings to json data

mitchute committed:
continued moving buildings to json data, revised building class

mitchute committed:
move to new buildings class

mitchute committed:
cleanups

mitchute committed:
cleanups

mitchute committed:
cleanup matrix generation

mitchute committed:
parallel components list

mitchute committed:
moving to unified component list

mitchute committed:
cleanout old code

mitchute committed:
cleanups

mitchute committed:
use topology object

axelstudios committed:
Fixed mypy errors

mitchute committed:
update schema

mitchute committed:
cleanups

mitchute committed:
update output headers

mitchute committed:
cleanups

mitchute committed:
fix bug where bh-resist was used instead of bh-radius

mitchute committed:
add another simple test, only load the objects contained in topology, commit output result diffs from de3bed755e45d6fca8c3c2e01f7c71f4f634aa78

mitchute committed:
rename vars

mitchute committed:
cleanup simple example, ensure all numpy array data types are float

mitchute committed:
testing simple network

mitchute committed:
Merge branch 'develop' into implement-sys-simulation

mitchute committed:
cleanup enum name

mitchute committed:
update schema

mitchute committed:
debugging

mitchute committed:
update test district loads

mitchute committed:
here we go again... starting to translate new code

mitchute committed:
cleanup output writing

mitchute committed:
move to json heat pump input data

mitchute committed:
cleanup, remove some unused stuff

mitchute committed:
eliminate a bunch of unnecessary array operations

mitchute committed:
moving fluid inputs to correct file

mitchute committed:
start walking GHE input data over to current json file, fix tests

mitchute committed:
move to use actual fluid props

mitchute committed:
cleanup

mitchute committed:
streamlining, cleanups

mitchute committed:
streamlining, cleanups

mitchute committed:
streamlining g-function calcs, removing unused buildings

mitchute committed:
streamline, cleanup

mitchute committed:
streamline, cleanup

mitchute committed:
print node indices

mitchute committed:
strip out all ISHX work

mitchute committed:
fix diffs in 1-pipe between old and new methods

mitchute committed:
add heat pump energy use

mitchute committed:
add central pump power

mitchute committed:
delete updated district implementation

mitchute committed:
cleanup

mitchute committed:
finish mapping district system simulation through input file

axelstudios committed:
Consistency improvements

mitchute committed:
fix file name

mitchute committed:
Merge branch 'develop' into implement-sys-simulation

mitchute committed:
fix merge issues

mitchute committed:
Convert some static functions to standalone functions

mitchute committed:
sizing_months to sizing_years

mitchute committed:
add multiyear simulations

mitchute committed:
add source-sink heat exchanger

mitchute committed:
fix incorrect temperature reporting in simulation, update app, update schema, fix tests

mitchute committed:
don't push to PyPI on pre-release

mitchute committed:
validate component names on topology, set minimum building flow rate, add another test

axelstudios committed:
Type fixes

axelstudios committed:
Fixed Python 3.10 issues

mitchute committed:
reduced digits in district output from 8 to 4

axelstudios committed:
Fix .python-version

TNWest committed:
Made changes to the manager class to allow for use in system sizing. (#164)

* Made changes to the manager class to allow for use in system sizing.

* Pin numpy to <2.4 to avoid x86-64-v2 errors on CI (#165)

* Pin numpy to <2.4 to avoid x86-64-v2 errors on CI

* Version bump

* Version bump

* Switch to managed python

* Bump uv

* make tests more robust

---------

Co-authored-by: Matt Mitchell <matt.mitchell@nrel.gov>

* Initial Sizing Algorithm Implemented.

* Iteration 2 of the system optimization method. Beginning work on the 3rd.

* Add hybrid timestep solver updates

* addition of hybrid timestep solver

* addition of hybrid timestep solver

* addition of hybrid timestep solver

* addition of hybrid timestep solver

* addition of hybrid timestep solver

* addition of hybrid timestep solver

* addition of hybrid timestep solver

* initial addition of horizontal code, updates to JSON and GHEHPsystem class to come

* Updated GHEHP system, added single pipe physics, added interpolation table creator, not yet tested

* Added the following: flag that can disable horizontal effects, variable discretization level in input file, capacitiy to handle parallel pipes, small interpolation table for running 3ghe-6bldg json example

Note: there are some pre-commit/linter errors, but I don't think they are a huge problem, so I'll fix them at a later date

* First completed version of the system design/sizing completed.

* Completed some minor changes and reformatting.

* Additional formatting fixes.

* Updated tests to account for new system simulation workflow.

* Updated schema and added ability to provide fixed COP rather than heat pump model for buildings.

* Updated schema and added ability to provide fixed COP rather than heat pump model for buildings.

* Flattened time_array to guarantee it is 1d.

* Changed guarantee to be the flatten method.

* checking new merged code (still in progress)

* fixed merge issues in horizontal code, next will do final verication to make sure code is outputting sensible values

* Horizontal code is outputting sensible values according to current tests, committing current version of code before attempting optimization

* optimized horizontal history calculation to vectorized version to avoid recalculating values

* Added TRT experimental verification for horizontal code (requires value tweaking to account for grout in test set-up)

* MInor review/changes to Two Pipe System equations based on discussion w/ NIranjan.

* Fixed formatting problems.

* fixed some formatting errors in horizontal code while testing

* removed unecessary JSON tests and fixed key error in horizontal interpolation

* patch tests

* patch test warnings

* format

* lint

* Bump urllib3 from 2.6.3 to 2.7.0 (#171)

Bumps [urllib3](https://github.com/urllib3/urllib3) from 2.6.3 to 2.7.0.
- [Release notes](https://github.com/urllib3/urllib3/releases)
- [Changelog](https://github.com/urllib3/urllib3/blob/main/CHANGES.rst)
- [Commits](https://github.com/urllib3/urllib3/compare/2.6.3...2.7.0)

---
updated-dependencies:
- dependency-name: urllib3
  dependency-version: 2.7.0
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

* Bump idna from 3.11 to 3.15 (#172)

Bumps [idna](https://github.com/kjd/idna) from 3.11 to 3.15.
- [Release notes](https://github.com/kjd/idna/releases)
- [Changelog](https://github.com/kjd/idna/blob/master/HISTORY.md)
- [Commits](https://github.com/kjd/idna/compare/v3.11...v3.15)

---
updated-dependencies:
- dependency-name: idna
  dependency-version: '3.15'
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

* Set min Python to 3.11 to use newer numpy (#166)

Co-authored-by: Matt Mitchell <matt.mitchell@nlr.gov>

* Address review issues

  - Moved unified_horizontal_library.pkl into ghedesigner/ghe/ and load it with importlib.resources, so
    horizontal simulations no longer depend on running from the repo root.

  - Initialized shared GHEHPSystem bookkeeping fields before branching on search_method, so
    SIMULATION_ONLY has the fields the CLI/output path expects.

  - Fixed the horizontal_piping schema to use real properties/required validation.
  - Added a dedicated horizontal_pipe schema because horizontal pipe inputs do not include borehole-only
    fields like shank_spacing.

  - Added regression tests for SIMULATION_ONLY initialization and invalid horizontal pipe schema
    rejection.

* update docs

* Address the following:

  - High: TWOPIPE matrix assembly leaves inlet_index as None for buildings and GHEs. See ghedesigner/
    district_system.py:1718 and use at ghedesigner/district_system.py:1176. NumPy treats row[None] = ...
    as a broadcasted assignment, so two-pipe matrices are silently corrupted rather than connected to a
    real inlet node.

* Address the following:

  - High: non-constant-COP TWOPIPE building simulation is unreachable because ghedesigner/
    district_system.py:1211 checks configuration == CentralLoopType instead of CentralLoopType.TWOPIPE.
    With constant_cop: false and TWOPIPE, Building.generate_matrix() returns None, so the solver will fail
    unpacking rows/RHS.

* Address the following:

  - Medium: fixed_loads building flow calculation is off by one and double-counts flow. ghedesigner/
    district_system.py:2171 computes one flow, adds it, then ghedesigner/district_system.py:2173 computes
    again with idx_timestep. calc_mass_flow_rate() indexes arrays directly at ghedesigner/
    district_system.py:1064, so the final timestep will index past the end, and earlier timesteps inflate
    loop flow.

* Address the following:

  - Medium: GroundHeatExchanger.get_g_function() accepts ghe_dict=None, but then leaves flow_rate and
    flow_type_str undefined unless a dictionary is passed. See ghedesigner/ghe/manager.py:716. The
    stateful manager path should use self.flow_rate / self.flow_type.

* remove logging, patch plotter app

* check results of other district sim tests

* remove duplicative example file

* auto load data files

* Path the following:

  - High: Non-constant-COP building simulation still uses the current unsolved timestep temperature when
    assembling the matrix. In ghedesigner/district_system.py:1203, self.t_in[idx_timestep - 1] reads the
    slot that will only be written after the solve at ghedesigner/district_system.py:2362. From timestep 2
    onward, COP coefficients are evaluated with the initialized ground temp instead of the previous solved
    inlet temp. This affects non-constant-COP ONEPIPE and TWOPIPE results silently.

* format

* Bump pytest from 9.0.2 to 9.0.3 (#173)

Bumps [pytest](https://github.com/pytest-dev/pytest) from 9.0.2 to 9.0.3.
- [Release notes](https://github.com/pytest-dev/pytest/releases)
- [Changelog](https://github.com/pytest-dev/pytest/blob/main/CHANGELOG.rst)
- [Commits](https://github.com/pytest-dev/pytest/compare/9.0.2...9.0.3)

---
updated-dependencies:
- dependency-name: pytest
  dependency-version: 9.0.3
  dependency-type: direct:development
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

* remove logger, loosen tolerance on tests

* looser still

* update diff app to better handle file paths

* clarify design pressure losses

* schema touchups, use a few constants

---------

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: Alex Swindler <alex.swindler@nrel.gov>
Co-authored-by: Matt Mitchell <matt.mitchell@nrel.gov>
Co-authored-by: TNWest <ultratigerofthewest@gmail.com>
Co-authored-by: niranjan-bas <niranjan.bastakoti@okstate.edu>
Co-authored-by: Drew McNair <drew.mcnair@okstate.edu>
Co-authored-by: Matt Mitchell <mitchute@gmail.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Matt Mitchell <matt.mitchell@nlr.gov>

mitchute committed:
Merge branch 'develop' into implement-sys-simulation

mitchute reviewed (commented):
All this has lived on side branches long enough. Let's merge and handle changes in a much smaller fashion going forward.
